src/field/FieldDriver.i.hh

```diff
@@ -0,0 +1,269 @@
+{
+    // Output with a step control error
+    FieldOutput output = this->find_next_chord(step, *state);
+
+    real_type step_taken = output.step_taken;
+
+    // Evaluate the relative error
+    real_type rel_error = output.value / (shared_.epsilon_step * step_taken);
+
+    if (rel_error > 1)
+    {
+        // Advance more accurately with a newly proposed step
+        real_type next_step = this->new_step_size(step, rel_error);
+        step_taken
+            = this->accurate_advance(step_taken, &output.state, next_step);
+    }
+
+    // Accept this accuracy and update the current state
+    *state = output.state;
+
+    return step_taken;
+}
+
+//---------------------------------------------------------------------------//
+/*!
+ * Find the next acceptable chord of which the miss-distance is smaller than
+ * a given reference (delta_chord) and evaluate the associated error.
+ */
+CELER_FUNCTION auto
+FieldDriver::find_next_chord(real_type step, const OdeState& state)
+    -> FieldOutput
+{
+    // Output with a step control error
+    FieldOutput output;
+
+    // Try with the proposed step
+    output.step_taken = step;
+
+    bool          converged       = false;
+    unsigned int  remaining_steps = shared_.max_nsteps;
+    StepperResult result;
+
+    do
+    {
+        result = stepper_(step, state);
+
+        // Check whether the distance to the chord is small than the reference
+        real_type dchord
+            = distance_chord(state, result.mid_state, result.end_state);
+
+        if (dchord <= shared_.delta_chord + rel_tolerance())
+        {
+            converged    = true;
+            output.value = truncation_error(
+                step, shared_.epsilon_rel_max, state, result.err_state);
+        }
+        else
+        {
+            // Estimate a new trial chord with a relative scale
+            output.step_taken
+                *= std::fmax(std::sqrt(shared_.delta_chord / dchord), half());
+        }
+    } while (!converged && (--remaining_steps > 0));
+
+    CELER_ASSERT(converged);
+
+    // Update new position and momentum
+    output.state = result.end_state;
+
+    return output;
+}
+
+//---------------------------------------------------------------------------//
+/*!
+ * Accurate_advance for an adaptive step control: Perform an adaptive step
+ * integration for a proposed step or a series of sub-steps within a required
+ * tolerance until the the accumulated curved path is equal to the input step
+ * length.
+ */
+CELER_FUNCTION real_type FieldDriver::accurate_advance(real_type step,
+                                                       OdeState* state,
+                                                       real_type hinitial)
+{
+    CELER_ASSERT(step > 0);
+
+    // Set an initial proposed step and evaluate the minimum threshold
+    real_type end_curve_length = step;
+
+    real_type h = ((hinitial > rel_tolerance() * step) && (hinitial < step))
```

Comment:
This might be a little confusing.  If I read correctly this seems to say "use hinitial if it not too large (larger than step) and not too small (smaller that some fraction of step), otherwise use step.   'Why?'  Is the wrong value of `hinitial` an exceptional condition (i.e. the caller should not have used this value) and the test should be turned into an assert or is it a routine occurrence?

Reply:
As a trial initial step,  hinitial is used if  hinitial < step (with not too small value of  hinitial with respect to step).  Otherwise, the use the full step length for the first trial.  Honestly, I also do not completely understand the logic behind, which I literally copied from Geant4 and modified little bit:
<pre>
  if ( (hinitial > 0.0) && (hinitial < hstep)
    && (hinitial > perMillion * hstep) )
  {
     h = hinitial;
  }
  else  //  Initial Step size "h" defaults to the full interval
  {
     h = hstep;
  }
</pre>

Reply:
Please add a comment.  We probably should review this more carefully at some point.

Reply:
Added a comment.